Updates for new Versioning Policy

Grauldon opened the pull request:
This PR addresses SpongeDocs #768 and [SpongeForge Issue #2367](https://github.com/SpongePowered/SpongeForge/issues/2367).

It changes SpongeDocs by creating a new directory for the topic and adding pages explaining how Sponge manages versions. It also updates existing files to move version topics to one place within the SpongeDocs contents and adds references from existing locations to the new pages. The home page is also updated to reference the new pages.

The first commit updates SpongeDocs to provide the current Java version update shipped with the Mojang installer (1.8.0_74). This information is on the Minecraft Wiki for the Windows versions; I downloaded the Mac OS X  version and confirmed 1.8.0_74 is inside the .dmg file.

Grauldon committed:
Updated Java version shipped with Mojang Installer inside of note

Grauldon committed:
Updated SpongeDocs for new versioning policy

phit commented on source/contributing/versioning.rst:
```diff
@@ -30,8 +13,8 @@ The Bleeding Branch
-new features, changes, and bugfixes. The version of ``bleeding`` will always be the next major release version
-appended with ``-SNAPSHOT`` (e.g. ``6.0.0-SNAPSHOT``) to denote that it is not yet a final build and subject to change.
+new features, changes, and bugfixes. The version of ``bleeding`` will always be the next major release version 
+appended with ``-SNAPSHOT`` (e.g. ``8.0-SNAPSHOT``) to denote that it is not yet a final build and subject to change.
```
should still be 8.0.0-SNAPSHOT

Grauldon replied:
Updated to 7.0.0-SNAPSHOT

phit commented on source/plugin/internals/implementation.rst:
```diff
@@ -27,7 +27,7 @@ Example Using Gradle
 
     dependencies {
-        compile 'org.spongepowered:spongecommon:7.1.0:dev'
+        compile 'org.spongepowered:spongecommon:8.1.0:dev'
```
why change these to versions that dont exist yet?

Grauldon replied:
Reverted version back to 7.1.0:dev.

phit commented on source/versions/filenames.rst:
```diff
@@ -0,0 +1,121 @@
+When you download files, the name of the file will provide some important version information. The following sections 
+describe the information provided.
+
+SpongeAPI
```
move spongeforge and vanilla above api

Grauldon replied:
Done

phit commented on source/versions/filenames.rst:
```diff
@@ -0,0 +1,121 @@
+
+.. note::
+
+    Sponge changed our versioning policy in October 2018. **Beginning with** SpongeAPI 8.0, the following information 
```
> Our versioning policy was updated in October 2018

Grauldon replied:
Changed

phit commented on source/versions/filenames.rst:
```diff
@@ -0,0 +1,121 @@
+| ``Version``          | When this number is **not** zero, the SpongeAPI version in development is used (i.e. API      |
+|                      | Minor + 1).                                                                                   |
++----------------------+-----------------------------------------------------------------------------------------------+
+| ``-RC``              | The build number in development for the next implementation release. When a build number      |
```
implementation > recommended

Grauldon replied:
Done

phit commented on source/versions/filenames.rst:
```diff
@@ -0,0 +1,121 @@
+| ``MinorRelease``     | :ref:`semantic versioning<sem-ver>`).                                                         |
++----------------------+-----------------------------------------------------------------------------------------------+
+| ``Recommended``      | The released version of the implementation when **not** followed by ``-RC<BuildNumber>.``     |
+| ``Version``          | When this number is **not** zero, the SpongeAPI version in development is used (i.e. API      |
```
I mean I guess, I would just leave this sentence out.

Grauldon replied:
Removed

phit commented on source/versions/filenames.rst:
```diff
@@ -0,0 +1,121 @@
+SpongeForge Jar files will always follow this naming scheme to allow you to easily identify compatibility.
+
+For example, the file name ``spongeforge-1.12.2-2705-7.1.4.jar`` is compatible with Minecraft version ``1.12.2``, was 
+built with Forge ``14.23.4.2705`` (Build ``2705``), uses SpongeAPI ``7.2-SNAPSHOT``, is a recommended version, and is 
```
> uses SpongeAPI ``7.2-SNAPSHOT`` 

should be

> guarantees compatibility with SpongeAPI ``7.1.0``

Grauldon replied:
Changed here and in the following example.

phit commented on source/versions/filenames.rst:
```diff
@@ -0,0 +1,121 @@
+
+The format of the filename is:
+
+``spongeapi-<APIMajor>.<LatestAPIMinorRelease>-shaded.jar``
```
``spongeapi-<APIMajor>.<LatestAPIMinorRelease>.0-shaded.jar``

Grauldon replied:
Done

phit commented on source/versions/filenames.rst:
```diff
@@ -0,0 +1,121 @@
+
+**OR**
+
+``spongeapi-<APIMajor>.<LatestAPIMinorRelease>.<Date>.<Time>-<BuildNumber>-shaded.jar``
```
``spongeapi-<APIMajor>.<NextAPIMinorRelease>.0-<Date>.<Time>-<BuildNumber>-shaded.jar``

Grauldon replied:
Done

phit commented on source/versions/filenames.rst:
```diff
@@ -0,0 +1,121 @@
+the ``4`` th release of SpongeForge with this API.
+
+Another example is the file name ``spongeforge-1.12.2-2705-7.1.5-RC3449.jar``. This file is compatible with Minecraft 
+version ``1.12.2``, was built with Forge ``14.23.4.2705`` (Build ``2705``), uses SpongeAPI ``7.2-SNAPSHOT``, is not a 
```
same as above for api

Grauldon replied:
Done

phit commented on source/versions/legacy-versions.rst:
```diff
@@ -0,0 +1,39 @@
+
+.. note::
+
+    Sponge changed our versioning policy in October 2018. **Prior to** SpongeAPI 8.0, the following information 
```
> Our versioning policy was updated in October 2018. **Prior to** SpongeAPI 8.0.0, ....

Grauldon replied:
Done

phit commented on source/versions/filenames.rst:
```diff
@@ -0,0 +1,121 @@
+    Sponge changed our versioning policy in October 2018. **Beginning with** SpongeAPI 8.0, the following information 
+    applies. See :doc:`legacy-versions` for information on older versions.
+
+When you download SpongeForge, the name of the file includes a Forge build number which this version of SpongeForge is 
```
I know this is old, but this isn't really true, would be nice if you can update this. The version in the filename is the version SpongeForge was built with, that guarantees compatibility, however most of the time newer Forge version also work fine but are not officially supported. So basically just word it differently, same applies to the same section on the old versioning page.

Grauldon replied:
Done

phit commented on source/versions/versioning.rst:
```diff
@@ -0,0 +1,118 @@
+developers. *SNAPSHOT* is associated with the API, and *RC*, or Release Candidate, is associated with the 
+implementation (SpongeVanilla and SpongeForge).
+
+*SNAPSHOT* and *RC* are **always** associated with a number (e.g., ``1-SNAPSHOT``, ``-RC120``). With *SNAPSHOT*, the 
```
``1-SNAPSHOT`` ?
what does that mean, snapshot is a suffix for the whole version before that, you cant just take the patch version by itself.
7.1.0-snapshot unlike RC120 is also not the name for a specific version, its refers to every build between releases.

phit commented on source/versions/versioning.rst:
```diff
@@ -0,0 +1,118 @@
+An important part to grasp in understanding Sponge versioning is how the version string conveys the API information. 
+The API version is primarily conveyed by the *major* . *minor* labels. The *major* label is always a number. The 
+*minor* label can be just a number, indicating a release, or it can be a number with ``-SNAPSHOT`` appended to it, 
+indicating a **minor release in development**. Developers are the most likely to use snapshots, so most Sponge users 
```
> Developers are the most likely to use snapshots

hopefully not, devs should use the latest stable release, snapshot features may break at any time if they arent part of the last release

Inscrutable replied:
It's often the way with other APIs, Devs will target new features they wanted, even if unstable, so SNAPSHOTS will still get used now and then. This is very much at their own risk, so development against a Recommended Build is preferred and encouraged.

Grauldon replied:
Updated

phit commented on source/versions/versioning.rst:
```diff
@@ -0,0 +1,118 @@
+SpongeAPI is the actual API that allows plugins to interact with the game. However, the API is enabled by the Sponge 
+implementation (SpongeForge or SpongeVanilla). Together, these products form the foundation of a Sponge server.
+
+As such, the version string reflects the state of these two products. The major and minor release components of a 
```
I don't agree with this, the implementation does derive part of its version from the API for clarity sake, however the versions are disconnected. `8.0.2` will never be a thing in API and the part derived from the API is `8.0.0` not `8.0`.

So a lot on this page does not make much sense to me personally, how it is explained and how you split the version string, but maybe I'm seeing it wrong. We will see what others say, just leave it for now.

Inscrutable commented on source/versions/filenames.rst:
```diff
@@ -0,0 +1,121 @@
+compatible with. Other builds, even ones differing by only a few build numbers, are not officially supported.
+
+However, SpongeForge usually updates to a new Forge build fairly soon after it's released, so you needn't worry about 
+always having to run an outdated Forge version in order to use SpongeForge.
```
I'm not sure this is really true, or useful (reflecting phit's comments above). When a Forge release breaks SpongeForge, we try to get an update out, but otherwise it's a bit ad-hoc. Normally updating Forge is a safe operation, but there is no particular rule I can think of that covers it nicely. Other than "try it and see what happens".

Grauldon replied:
I used phit's comment to rewrite those paragraphs. We may need to make more changes after you review the new update.

Grauldon replied:
I realized you were mentioning the paragraph after phit's comment. So, I re-worked the changes to handle both comments. We still may need to make further changes, but I think you will like the new changes.

Inscrutable commented on source/versions/versioning.rst:
```diff
@@ -0,0 +1,118 @@
+this version string as an example: ``8.2.0``. This string conveys the API is Version 8.2 and the implementation is the 
+first release using this API version.
+
+The following table provides several examples of a version string and its components:
```
This table does not help much - it may actually add to confusion. I'm not sure dissecting the semver this way is really assisting. Maybe directing via to a use-case might improve it.

ImMorpheus replied:
What about linking to the website like github does ? 
https://semver.org/

Grauldon replied:
Changed

Inscrutable reviewed (commented):
Excellent. I'll see what others have to add to this. We want it to be concise.

ImMorpheus commented:
@Grauldon Thanks for your contribution.

Just a note: the stable branch is used for the [7.X.Y version of the docs](https://docs.spongepowered.org/stable/en/).
For consistency and to avoid confusion once there's a dedicated docs for API 8, it's better to keep any version reference to API 7. Especially in `Example Using Gradle`.

Grauldon commented:
I'll plan to return to this PR this week. My original plan was to balance time between this one and #784 better. Using git stash to rebase with stable has helped me to realize how to accomplish this. Thanks for all of the feedback!

Grauldon committed:
Updated per comments in PR

Grauldon committed:
Reverted version per comments in PR

Grauldon committed:
Made updates and corrections to filename docs per comments in PR

dualspiral commented on source/versions/versioning.rst:
```diff
@@ -0,0 +1,118 @@
+==========
+Versioning
```
First, please let me thank you for all the work you've done so far. I do not want to detract from it and this thought process would not exist if you hadn't stepped up, so thank you!

However, this particular page doesn't feel like it's particularly useful in this form. We talk about sem-ver, what it means, then the version string, but because there is so much on here, when you look at the API version string, for example, people may have lost the link to sem ver. It is better to explain what each bit means inline for us.

I think what I say needs discussion (@SpongePowered/docs) and this is just my own opinion, but I would:

* Remove the semver section, with the exception of a tip at the top saying "Sponge follows a format of the SemVer specification in its projects. You can read about general SemVer usage at http://semver.org". There really is no point in us rehashing it.
* Say that SpongeAPI and SpongeForge/SpongeVanilla follow two different versioning policies
* Highlight the important parts of the version string as the semver part and, for impl, the minecraft version
* Explain what that means for the developer (for the API version string) and the server owner (for the impl version string) - most of this page doesn't really decribe that and that's what the versioning is really all about for us.

For the API, it needs to boil down to this:

* A change in the major number indicates that some APIs that were guaranteed in the previous version have been broken. 5->6 broken world pregen APIs, for example. 6->7 was a more significant break. A bump in the major number is always accompanied by resettign the minor number to zero.
  * Developers know that their plugins _might_ break when this changes, i.e. API 6.x plugins might not run on API 7.x.
* A change in the minor number (but not the major number) means that API calls have been added, but plugins for previous minor revisions of the same major revision will still work - _but this isn't necessarily the case the other way around_.
  * Developers know their plugins will still function on 7.x+n (target 7.x, n≥0, so 7.0 plugins work on 7.1), but not necessarily the other way around.
* `-SNAPSHOT` means that the versioned API is still in progress
  * If the minor version is 0 (e.g. 8.0-SNAPSHOT), **anything** may break before it is released and plugins cannot expect stability.
  * If the minor version is greater than zero (e.g. 7.2-SNAPSHOT), anything up to the previous minor revision is still guaranteed (in this case, 7.1), but anything added since (in this case, for what will be 7.2) **may still break**, so cannot expect stability on unreleased features.

For the impl:

* A version string with an RC identifier means "recommended build", builds we consider as a reasonable quality that we would want to run ourselves.
* We tend to bump the build to indicate bug fixes and performance improvements, or any significant changes in the config or other systems that aren't related to the API (in this sense, the "patch" part of the version string isn't really a "patch" part and so it kind of moves away from the sem-ver ideology).
* For the 7.x.y part, 7.x is the version of the API that we guarantee for plugins, y is the recommended build version for this API series.
* Plugins written for 7.0 to 7.x will work. 7.x+n (n≥1) (including any SNAPSHOT features) _may not_ work.
* Any server software that has the version `n.0.0-SNAPSHOT` where the minor and recommended build numbers are 0 (or would the recommended build release be 1, I'd have to check) and the SNAPSHOT monkier is applied, versions are bleeding builds for testing the next version of the API only and may break at any time as we refine everything about it.

Grauldon replied:
Thank you for the comment and feedback! This in-depth explanation provides an outline I can follow to get it right this time. I figured a re-write was needed; I just wasn't sure yet how to proceed. Your feedback will save me much time. Thanks again!

dualspiral replied:
To be honest, what you had wasn't bad, it just seemed to kind of meander a bit much. You were doing a good job of being descriptive, just in the wrong areas! Regardless, I know that large comments that suggest a rewrite can be hard to take, so thank you for taking it in the good faith you have! :)

Grauldon committed:
Missed two corrections of version in filename

Grauldon commented:
I have updated the PR to address most of the reviews provided. Thanks to all for the feedback and pointing out mistakes and better ways of writing the documentation.

I did not work with versioning.rst. I had planned on spending much more time with it, and I have run out of time today. So, I am pushing the other changes I have made and will continue with the versioning.rst file Saturday. 

@ImMorpheus Thank you for the comment and feedback. The link is near the top. But like @dualspiral pointed out, the deluge of information drowns it out. Also, I made the corrections to `Example Using Gradle`. 

I will be away from my computer tomorrow, so I won't be able to make any updates until Saturday. I should be able to look over the feedback about versioning.rst some tomorrow though. I will push those changes Saturday evening if possible.

ImMorpheus commented on source/versions/filenames.rst:
```diff
@@ -0,0 +1,117 @@
+
+When you download SpongeForge, the name of the file includes a Forge build number which this version of SpongeForge is 
+built with and guaranteed to be compatible. We do not support newer versions of Forge; however, we try to get an update 
+out when a Forge release breaks SpongeForge. You can try the Forge version needed as it will work fine most of the 
```
This isn't true either.
SpongeForge is built against the latest **recommended** version of Forge, we don't bump the Forge version unless there's a new recommended build. 

See https://github.com/SpongePowered/SpongeForge/issues/2416 https://github.com/SpongePowered/SpongeForge/issues/2370

dualspiral replied:
That said, we've also discussed updating to non-recommended builds, while it doesn't happen now, we should not be boxing ourselves into such a policy in future (and we should avoid making guarantees to something like this), especially as we will probably not think to update a sentence like this after the fact.

I would just delete the sentences "We do not support... breaks SpongeForge" and "If it doesn't, we... version soon" (but consider my other notes).

Grauldon replied:
Done

ImMorpheus commented on source/versions/filenames.rst:
```diff
@@ -0,0 +1,117 @@
+    When investigating crash issues, you can freely try newer versions of Forge than listed on the SpongeForge Jar.
+    However, it is recommended to also check with the matching version, to make sure your issue is not related to a
+    version mismatch.
+    Even though there will be no guarantee of compatibility, please report any breakage to the issue tracker, so that
```
See above.

This is misleading, if you're not using the recommended forge version then there's nothing to fix as we won't break compatibility with the recommended build.

dualspiral replied:
I agree with @ImMorpheus (with my caveats, above), [but I just want to point out this was a copy/paste](https://github.com/SpongePowered/SpongeDocs/pull/785/files/3925d9afa55f368d4d6300331e443d9da05e4f19#diff-1b21b02d7e6287095c0a769eab09ebc1L64). This is a repeat of what's already been said anyway, so this entire warning could probably be removed.

Grauldon replied:
Warning removed; note handled in accordance with [comments below](https://github.com/SpongePowered/SpongeDocs/pull/785#discussion_r261731618).

dualspiral commented on source/versions/filenames.rst:
```diff
@@ -0,0 +1,117 @@
+
+.. note::
+
+    Our versioning policy was updated in October 2018. **Beginning with** SpongeAPI 8.0, the following information 
```
We started this with Sponge API 7.1+ builds

Grauldon replied:
Corrected

dualspiral commented on source/versions/filenames.rst:
```diff
@@ -0,0 +1,117 @@
++----------------------+-----------------------------------------------------------------------------------------------+
+| ``MCVersion``        | The Minecraft version. Only clients compatible with this version can connect.                 |
++----------------------+-----------------------------------------------------------------------------------------------+
+| ``ForgeBuild``       | Preferably your server should run this exact version of Forge (which can be identified in the |
```
"The Forge build that SpongeForge is built against and is guaranteed to be compatible with". Describe what this means outside of the table.

Grauldon replied:
Changed

dualspiral commented on source/versions/filenames.rst:
```diff
@@ -0,0 +1,117 @@
+| ``-RC``              | The build number in development for the next recommended release. When a build number is      |
+| ``<BuildNumber>``    | present, the ``RecommendedVersion`` has **not** been released yet.                            |
++----------------------+-----------------------------------------------------------------------------------------------+
+
```
I would move the description of the significance of the forge build number here. You can then say:

* SpongeForge will work with the Forge build ``ForgeBuild``
* SF may work with other Forge versions, but we don't guarantee this
* We **tend** to compile against the latest Recommended Build for Forge

The note, warning and description are too far away from one another, bring them all together here (though it was an artifact of what you started with!). Show what the version string is, then start talking about its significance, not the other way around.

Grauldon replied:
Removed warning, moved the note to below table and described Forge build between the table and the note. Copied changes to the legacy-versions.rst file (except I excluded "**tend** to" in statement about Forge's Recommended Build).

dualspiral commented on source/versions/legacy-versions.rst:
```diff
@@ -0,0 +1,37 @@
+
+.. note::
+
+    Our versioning policy was updated in October 2018. **Prior to** SpongeAPI 8.0, the following information applies. 
```
8.0 -> 7.1

Grauldon replied:
Corrected

dualspiral reviewed (commented):
I'm not convinced about splitting into versions and filenames, as the filename for our impls is pretty much the version. I would perhaps split it into API and Implementation versions and rejig stuff so server owners and plugin developers have targetted pages. It doesn't invalidate anything that's happened here as most of what is written may just need a cut and paste job into different homes.

So, something like this:

```
                                  |--> SpongeAPI Versioning & Filenames
Versioning (inc. common stuff) ---|--> Legacy SF/V Versioning
                                  |--> SpongeForge/Vanilla Versioning & Filenames
```

I would hold off on doing this until that's discussed though - **this isn't a request for change but a request for input** to try to ensure that duplication is minimised and that people can minimise the pages they look at (the quick start guide for setting up a server came about because we had a lot of pages for how to set up a server, but no "reference page").

It would be instructive, should there be any confusion, to read https://github.com/SpongePowered/SpongeForge/issues/2367 where the version changes were proposed which might give clarity.

Else, I have a few comments below.

gabizou commented on source/versions/filenames.rst:
```diff
@@ -0,0 +1,117 @@
+.. note::
+
+    The first format without the build information is the release and recommended build format. The second format with 
+    build information is the *SNAPSHOT* format. This version is still in development.
```
Maybe mention that this format is only used as raw downloads. Depending on the API via Gradle or Maven will download a formatted named jar as the recommended build format (`spongeapi-7.2.0-SNAPSHOT.jar`). Downloading via browser for specific builds will result in the timestamped jar file.

Grauldon replied:
`spongeapi-7.2.0-SNAPSHOT.jar` is the recommended build format? I thought the `-SNAPSHOT` label means it is still in development.

Grauldon replied:
Added another format. May need to change further to have the appropriate label.

Grauldon committed:
Rewrote versioning.rst and corrected/improved filenames.rst

Grauldon commented:
I rewrote versioning.rst and corrected/improved filenames.rst.

>  I would perhaps split it into API and Implementation versions and rejig stuff so server owners and plugin developers have targetted pages.

@dualspiral I get what you are saying, and I agree, although I had already made some changes along these lines. I left the changes I had made to see if they suffice and to allow input from others. It wouldn't take much time to make the changes, so we can go that route if decided. I think your solution might be better since I am concerned that readers will brush over the warning and read/use the balance of the page anyways.

Also, I have read #2367 quite a bit; actually, I have printed sections of it and highlighted parts throughout the pages. I am concerned this may be tripping me up though. I read all comments and try to implement them with the explanations in that issue as backdrops. My concern is that some of the information may have changed, but I keep stumbling because I am not aware of those changes. I don't want to come across as not listening or being hard headed. I'm just trying to use feedback in light of the explanations in the issue.

@gabizou Please check comments concerning the SpongeAPI filename format.

Thank you all for the comments and feedback!

phit commented on source/versions/filenames.rst:
```diff
@@ -57,22 +63,6 @@ version ``1.12.2``, was built with Forge ``14.23.4.2705`` (Build ``2705``), guar
-    use more recent versions of Forge, than the one used for SpongeForge, but we can't always guarantee compatibility.
-
-.. warning::
-
```
this still applies, why did you remove it?

Grauldon replied:
Oh, snap! At first, I saw it as a duplication with the new content. Had I seen the issue tracker part, I would have misunderstood a previous comment to mean not to use the issue tracker. I will put it back in.

Grauldon replied:
Corrected.

phit commented on source/versions/versioning.rst:
```diff
@@ -4,115 +4,111 @@ Versioning
+.. warning::
+
+    The following information is for Sponge Developers. Plugin Developers and Server Administrators should only use 
+    **Recommended Versions** for development and server installations. Doing otherwise is at your own risk and **not** 
```
sounds a bit too harsh, but I guess this is fine?

Grauldon replied:
How about "Using a version that is not a recommended build introduces instability and nearly guarantees problems. We strongly discourage the use of non-recommended builds by plugin developers and server administrators."?

Other options are just removing the last sentence or using dualspiral's idea of separate pages.

Grauldon replied:
Changed to statements above.

phit commented on source/versions/versioning.rst:
```diff
@@ -4,115 +4,111 @@ Versioning
+=================
+
+Release Candidates (**RC**) are builds in which bug fixes are released for implementations while they are still in 
+development. They are nearly ready for a full release, but require more bug fixes and testing.
```
as we run rolling releases, every commit is a RC!
they can range from not even launching to just as stable as a rb

Grauldon replied:
I don't understand what you're saying. Should I remove this section, add a statement about the rolling releases, or am I missing the release candidate idea completely?

phit replied:
Yes, anything that isn't an RB is an RC, it isn't just stuff that isn't ready, it's everything that gets pushed to Github doesn't matter what state it is in. Any RC can become an RB: e.g. these 3 builds are identical from the codebase https://i.imgur.com/wkClA3j.png

Grauldon committed:
Added warning removed by error, toned down another warning, and corrected RC section

Inscrutable commented on source/versions/versioning.rst:
```diff
@@ -104,9 +105,11 @@ A release which has the ``-SNAPSHOT`` label represents the next API version in d
+recommended build. Furthermore, the code may or may not change; the change might occur in a settings file. The 
+resulting build may run as well as a recommended build, or it may not run at all. In the end, a change occurred in the 
+repository, it triggered a build, and the resulting build is not a recommended build. Therefore, it is a release 
+candidate.
```
Brick-a-budgie, that's horrible to read. Accurate but rambling, methinks it has too many words by far. I mean, the only thing we really need to get across is that RC's *might not work properly*, but may become RB's after further testing. Getting into the nuts-and-bolts of the process (e.g. "triggered a build") is unnecessary, imho.

Grauldon committed:
Reworded release candidate section

Grauldon committed:
Converted relative path to absolute path

Signed-off-by: Grauldon <17791090+Grauldon@users.noreply.github.com>

Grauldon committed:
Added required changes missed during earlier review

Signed-off-by: Grauldon <17791090+Grauldon@users.noreply.github.com>

Grauldon commented:
Added two more commits:

1. Found another case where a relative path was more than two deep
2. Found changes that should have occurred with the original movement of the filename explanation

Spongy commented:
A preview for this pull request is available at https://cdn.rawgit.com/Spongy/SpongeDocs-PRs/b9c4ad9/.

Here are some links to the pages that were modified:

- added: https://cdn.rawgit.com/Spongy/SpongeDocs-PRs/b9c4ad9/versions/filenames.html
- added: https://cdn.rawgit.com/Spongy/SpongeDocs-PRs/b9c4ad9/versions/index.html
- added: https://cdn.rawgit.com/Spongy/SpongeDocs-PRs/b9c4ad9/versions/legacy-versions.html
- added: https://cdn.rawgit.com/Spongy/SpongeDocs-PRs/b9c4ad9/versions/versioning.html
- modified: https://cdn.rawgit.com/Spongy/SpongeDocs-PRs/b9c4ad9/contributing/versioning.html
- modified: https://cdn.rawgit.com/Spongy/SpongeDocs-PRs/b9c4ad9/index.html
- modified: https://cdn.rawgit.com/Spongy/SpongeDocs-PRs/b9c4ad9/plugin/api-versions.html
- modified: https://cdn.rawgit.com/Spongy/SpongeDocs-PRs/b9c4ad9/plugin/internals/implementation.html
- modified: https://cdn.rawgit.com/Spongy/SpongeDocs-PRs/b9c4ad9/server/getting-started/implementations/spongeforge.html
- modified: https://cdn.rawgit.com/Spongy/SpongeDocs-PRs/b9c4ad9/server/getting-started/implementations/spongevanilla.html
- modified: https://cdn.rawgit.com/Spongy/SpongeDocs-PRs/b9c4ad9/server/spongineer/debugging.html
- modified: https://cdn.rawgit.com/Spongy/SpongeDocs-PRs/b9c4ad9/server/spongineer/logs.html

_Since the preview frequently changes, please link to [this comment](https://github.com/SpongePowered/SpongeDocs/pull/785#issuecomment-471165504), not to the direct url to the preview._

ST-DDT commented on source/plugin/api-versions.rst:
```diff
@@ -37,41 +40,7 @@ Getting the API Version from Implementations
-
-    Major API version bumps will be reserved for the experimental / bleeding builds, and will be very explicit as their
-    major version number will be different.
+See :doc:`../../versions/versioning`.
```
Use absolute paths here.

Grauldon replied:
Changed

ST-DDT commented on source/plugin/internals/implementation.rst:
```diff
@@ -18,7 +18,7 @@ SpongeCommon
 - **Group ID**: ``org.spongepowered``
 - **Artifact ID**: ``spongecommon``
-- **Version**: Same as SpongeAPI
+- **Version**: Same as SpongeAPI + current recommended version (see :doc:`../../../versions/versioning` for details)
```
and here

Grauldon replied:
Changed

ST-DDT commented on source/server/spongineer/debugging.rst:
```diff
@@ -23,7 +23,7 @@ Is the recommended Forge version installed?
 However, it is strongly advised to run the recommended build only.
 If you encounter a crash and your versions are mismatching, match them first and try again.
-If you're unsure which Forge build you need, take a look at :doc:`/server/getting-started/implementations/spongeforge`
+If you're unsure which Forge build you need, take a look at the **SpongeForge** section of :doc:`../../versions/filenames`.
```
and here

Grauldon replied:
Changed

ST-DDT commented on source/server/spongineer/debugging.rst:
```diff
@@ -95,7 +95,7 @@ Mismatched SpongeForge and Forge
 This is a common crash when you try to run SpongeForge on the wrong Forge build. Note that the target/Mixin can vary.
 Always match Forge against SpongeForge! If you're unsure which version of Forge is required and you already got your
-SpongeForge build, take a look at: :doc:`../getting-started/implementations/spongeforge/`
+SpongeForge build, take a look at the **SpongeForge** section of :doc:`../../versions/filenames`. 
```
and here

Grauldon replied:
Changed

ST-DDT commented on source/server/spongineer/logs.rst:
```diff
@@ -182,7 +182,7 @@ The example log indicates that we're running:
 
 This indicates that SpongeForge 3136 was found and loaded by Forge. For further help regarding the SpongeForge
-naming scheme, have a look here: :doc:`../getting-started/implementations/spongeforge/`.
+naming scheme, have a look at :doc:`../../versions/filenames`.
```
and here

Grauldon replied:
Changed

ST-DDT reviewed (commented):
Nice changes!

Grauldon committed:
Converted relative paths to absolute paths

Signed-off-by: Grauldon <17791090+Grauldon@users.noreply.github.com>

Inscrutable commented:
@ImMorpheus - your review is the last hurdle. If you're happy with the changes we can publish this.

ImMorpheus commented on source/plugin/internals/implementation.rst:
```diff
@@ -46,7 +46,7 @@ Example Using Gradle
 
     dependencies {
-        compile 'org.spongepowered:spongevanilla:1.12.2-7.1.0-BETA-116:dev'
+        compile 'org.spongepowered:spongevanilla:1.12.2-7.1.0-RC116:dev'
```
This won't compile.

The version change happened with RC123. 
There's no spongevanilla:1.12.2-7.1.0-RC116.

See https://www.spongepowered.org/downloads/spongevanilla/stable/1.12.2

Grauldon replied:
Corrected

ImMorpheus commented on source/plugin/internals/implementation.rst:
```diff
@@ -65,5 +65,5 @@ Example Using Gradle
 
     dependencies {
-        compile 'org.spongepowered:spongeforge:1.12.2-2705-7.1.0-BETA-3442:dev'
+        compile 'org.spongepowered:spongeforge:1.12.2-2705-7.1.0-RC3442:dev'
```
Same as above.

The versioning change for SF happened with RC3484

Grauldon replied:
Corrected

ImMorpheus commented on source/versions/filenames.rst:
```diff
@@ -0,0 +1,126 @@
+built with Forge ``14.23.4.2705`` (Build ``2705``), guarantees compatibility with SpongeAPI ``7.1.0``, is a recommended 
+version, and is the ``4`` th release of SpongeForge with this API.
+
+Another example is the file name ``spongeforge-1.12.2-2705-7.1.5-RC3449.jar``. This file is compatible with Minecraft 
```
There's no spongeforge-1.12.2-2705-7.1.5-RC3449, but this is used as an example so I'm not sure it's worth changing

Grauldon replied:
I agree with using correct filenames, so I went ahead and corrected them. I also updated the SpongeVanilla example as well.

ImMorpheus commented on source/versions/legacy-versions.rst:
```diff
@@ -0,0 +1,51 @@
+
+SpongeForge Jar files will always follow this naming scheme, to allow you to easily identify compatibility.
+
+For example, the file name ``spongeforge-1.12.2-2705-7.1.0-BETA-3442.jar`` is compatible with Minecraft version
```
See above

Grauldon replied:
This filename was released (see [SpongeForge download](https://www.spongepowered.org/downloads/spongeforge/stable/1.12.2?until=2018-09-14T15%3A28%3A16Z)), so I left it.

Grauldon committed:
Corrected examples to use accurate filenames

Signed-off-by: Grauldon <17791090+Grauldon@users.noreply.github.com>

ST-DDT commented on source/versions/filenames.rst:
```diff
@@ -0,0 +1,126 @@
+
+``spongeforge-<MCVersion>-<ForgeBuild>-<APIMajor>.<LatestAPIMinorRelease>.<RecommendedVersion(-RC<BuildNumber>)>.jar``
+
++----------------------+-----------------------------------------------------------------------------------------------+
```
I'm not sure, but the tables without headers look a little bit strange. Should we add the headers?

ST-DDT reviewed (approved):
I'm unsure about the missing table headers, but otherwise it looks good to me.
Preview updated.

dualspiral reviewed (approved):
I'll add my review! :) 

I think this is good to go. We can fix minor quibbles later as they come up, no point batting this around further.

phit commented:
should versioning be a main category? if it should it should be moved down in the sidebar

ST-DDT commented:
> should versioning be a main category? 

Yes, unless you have a good idea where to place it otherwise.

> it should be moved down in the sidebar

I somewhat agree with that. Any suggestion for the target ordering?

phit commented:
probably like this, so after creating a server and before dev work as kinda a bridge

    Creating a Server
    Versioning Policy
    Preparing for Development
    Creating a Plugin
    Ore Documentation
    Contributing to Sponge
    About the Sponge Project

the ordering and naming kinda needs a general makeover anyway but thats out of scope for this pr

Grauldon commented:
I like phit's suggestion on the sidebar. I can also add the table headers while making that update, perhaps `Component` for the first column and `Purpose` for the second. 

I also realized I forgot a parenthetical note on `Recommended Version` for the SpongeForge table while looking at the latest comments. I want to update it from 

> The released version of the implementation when **not** followed by ``-RC.``

to

> The released version of the implementation when **not** followed by ``-RC`` (the ``Z`` in :ref:`semantic versioning`). 

It will look just like the parenthetical notes in ``APIMajor`` and ``LatestAPIMinorRelease`` (except for stating Z), and it will complement Sponge's use of SemVer.

Any thoughts on table headers and the note or further comments about phit's suggestion?

Inscrutable commented:
Those headless tables seem OK by me, they are just another way of doing a list. 
I agree with dualspiral. 🏏 Add the note, tweak the order, and it can be finished.

Grauldon committed:
Reorder the sidebar and add parenthetical note for SemVer

Signed-off-by: Grauldon <17791090+Grauldon@users.noreply.github.com>
